### DISCUSSION TITLE
System Monitor and Spectrom not working with M10 or F10 receivers.

### DISCUSSION THREAD

**Discussion by GPSFan** [semuconsulting/PyGPSClient#125] [General]
I am using the latest PyGPSClient, and both System Monitor and Spectrum work with my F9 receivers, but not with the M10 or F10. Looking at the protocol specs for both parts, it seems like the commands to turn on the SPAN data are the same. II have tried to turn on UBX-MON-SPAN and UBX-MON-COMMS via the UBX config interface and while I get an ACK for those commands, I don't get any COMMS or SPAN data.
SPAN works fine on u-center with all the receivers I have.

**Comment by semuadmin** [comment]
Hi @GPSFan 

Which specific M10 / F10 devices are you using? If the interface specifications for those specific devices include MON-SPAN and MON-COMMS messages, there should be no reason why the spectrum and system monitor widgets shouldn’t work.

Are you seeing incoming MON-SPAN or MON-COMMS messages in the console?

**Reply by GPSFan** [reply]
I have a Sparkfun breakout board for the M10S and one for the Neo-F10N. They are real u-blox parts not something a CN Aliblabla seller salvaged. Teh M10S has firmware 5.10 and the F10 has 6.00.
When I use u-center to turn on MON-SPAN I see the binary messages in the binary packet view. When I try the same thing with PyGPSClient I get an ACK for the command but don't see any messages. When I use my F9P board PyGPSClient turns on the MON-SPAN just fine and I see the spectrum display.
The protocol specs for the M10 and F10 firmwares 5.1 and 6.0 are word for word the same as for the F9P (for the MON_SPAN and MON-COMMS commands. I have been using u-blox parts for quite some time and had some of the first F9P parts in the US. I built my own PCB for those parts.

**Comment by semuadmin** [comment]
Ok so just to cover off the basics…

1. What version of PyGPSClient are you using? Is it the latest?
2. What platform (device/operating system) are you running on?
3. Do you have any other processes running on this platform that might be interfering with serial comms e.g. gpsd daemon?
4. Which physical serial port are you connecting to? Are you sure outbound UBX protocol is enabled on this port? Are you seeing _any_ UBX messages at the PyGPSClient console (other than ACK) e.g. standard NAV messages?
5. How, specifically, are you enabling the relevant UBX messages? Via PyGPSClients’s automatic ‘widget’ functionality, or via one or more manual CFG message(s)?
6. Are you using a PyGPSClient json configuration file? If so, what is the value of the ”defaultport_s” setting in that file (_this determines which port(s) any CFG-MSG commands apply to by default_)?
7. Are you running at an exceptionally high navigation rate (e.g. 10Hz or higher) and/or an exceptionally low baud rate (e.g. 9600)? _UBX can occasionally drop CFG configuration messages if the rx or tx buffers are full_
8. Have you tried factory-resetting the receivers?

**Reply by GPSFan** [reply]
1: 1.4.13 updated today
2: Linux Mint 21.3
3: Perhaps, let me shut them down and try again...No change..
4: /dev/ttyUSB0, 38400, yes,yes
5: both
6: no
7: no I have done reset to defaults and the rates are all 1hz
8: yes

**Reply by GPSFan** [reply]
What I would normally do at this time is set up a serial monitor to watch exactly what is being sent to the GNSS module and what it is saying in return, but today/yesterday/tomorrow and tomorrow are chock full of other things to do, so I thought I'd submit a bug report.

**Reply by semuadmin** [reply]
Sensible approach. I was going to ask if you could record a binary datalog in PyGPSClient covering the period when you’re enabling and disabling the spectrum and monitor widgets. 

There’s no functionality in PyGPSClient or its underlying parsers which discriminates between different receiver models, and it has been tested extensively with Sparkfun Gen10 UBX boards, so I’m at a bit of a loss to explain why it would work with an F9P but not an F10 in your situation. 


if you’re not seeing any MON-SPAN or MON-COMMS messages in the console, then the problem is clearly not with PyGPSClient’s “widgets” themselves, but with the devices’ response to a standard CFG-MSG or CFG-VALSET configuration message. When I get a moment I’ll double check the configuration database hex key values for those commands and see if they’ve changed in the latest M10 or F10 firmware, but that’s a long shot.

**Reply by semuadmin** [reply]
Not sure if you’ve tried this already, but try enabling MON-SPAN messages on your devices’ USB port by using the CFG-VALSET command with a key of CFG_MSGOUT_UBX_MON_SPAN_USB (0x2091038E), and then see if the spectrum widget works. It may be that the latest Gen10 firmware no longer supports CFG-MSG commands (they’ve been deprecated since Gen9), in which case it’s a simple fix.

**Reply by GPSFan** [reply]
I'll give it a try, not sure when I'll get time though. Maybe tomorrow. 
Let me pull out an M9 and see if displays the spectrum...
M9 works just fine. 
Is there a way to show the commands sent in PyGPSClient?

**Reply by semuadmin** [reply]
Yeah, you can ‘record’ any commands sent via the PyGPSClient UBX configuration panel using the record facility, and then play them back in the console. I’m away from my usual testing resources at the moment but I’m guessing this is about UBX-CFG-MSG being dropped in the latest Gen10 firmware. If so, it’s an easy fix.

**Comment by GPSFan** [comment]
Here is what my serial monitor sees for PyGPSClient and u-center 23.08:

command sent by PyGPSClient when span enabled ie click on the View menu Show Spectrum.

b5 62 06 01 08 00 0a 31 00 00 00 01 00 00 4b 4d // enable MON-SPAN at 1hz UART1
b5 62 06 01 08 00 0a 39 00 00 00 00 00 00 52 82 // disable NAV-GEOFENCE
b5 62 06 01 08 00 0a 36 00 00 00 00 00 00 4f 6d // disable MON-COMMS
b5 62 06 01 08 00 01 3c 00 00 00 00 00 00 4c 4f // disable NAV-RELPOSNED

Response from F10N
b5 62 05 01 02 00 06 01 0f 38 // ACK
b5 62 05 01 02 00 06 01 0f 38 // ACK
b5 62 05 01 02 00 06 01 0f 38 // ACK
b5 62 05 00 02 00 06 01 0e 33 // NAK!!

What u-center sends when I enable the MON-SPAN message in the Message View

b5 62 0a 31 00 00 3b bb // poll
Or
b5 62 06 01 03 00 0a 31 01 46 c0 // enable 1sec rate

There Seems to be a NAK for the last command sent by PyGPSClient.

**Reply by GPSFan** [reply]
For an F10N,
I tried using the CUSTOM command of u-center sending:
b5 62 06 01 08 00 0a 31 00 00 00 01 00 00 4b 4d
from u-center and got an ACK, but no SPAN messages, I then sent:
b5 62 06 01 03 00 0a 31 01 46 c0
and got a stream of SPAN messages, so while the 8 byte CFG-MSG command has been deprecated in protocol 40 it still gets an ACK but it doesn't do anything. while the 3 byte version still works.

**Comment by GPSFan** [comment]
When I enable the Spectrum and then go to the ubx config screen and then use VALSET to set the CFG_MSGOUT_UBX_MON_SPAN_UART1 to 1 I see that the Spectrum display now shows the proper spectrum as did my F9 and M9 modules.
But I only seem to get 1 message, not a stream. Correction, I do get a stream... But the display only seems to update once.

**Reply by semuadmin** [reply]
Many thanks for the investigation & feedback. Not sure why you’d only see one update - the display should refresh on receipt of each MON-SPAN. I’ll be able to get into this in detail next week.

**Reply by GPSFan** [reply]
Sounds good, no real rush, I want to do some antenna testing and have heard that MON-SPAN can do a quick and dirty eval of an L1/L2/L5 antenna.

**Reply by semuadmin** [reply]
FYI I've spun a new branch up [RC 1.4.14](https://github.com/semuconsulting/PyGPSClient/tree/RC-1.4.14), which includes a provisional fix for this issue.

Essentially I'm polling the ROM version of the device on connecting and, if it's greater than "23.01" (_the version in which the configuration database was introduced_), I use the newer CFG-VALSET command rather than the older CFG-MSG command to set message rates for MON-SPAN, MON-COMMS, etc.

**NOTE**: the device port(s) to which any message rate command is applied are defined by the `defaultport_s` configuration parameter. This can now be set via checkboxes in the UBX Configuration Preset panel:

<img width="384" alt="Screenshot 2024-04-09 at 13 50 06" src="https://github.com/semuconsulting/PyGPSClient/assets/28569967/6f957ded-ab0d-4beb-8532-2bf7d6afe500">

**Ensure that your selection is consistent with the device's capabilities**, otherwise any CFG-MSG or CFG-VALSET command may be rejected and the widget won't work - i.e. don't select `USB` if the device doesn't have a USB port (_as is the case with my SparkFun M10S module, for example_). It's done this way as there does not appear to be any 100% foolproof way of detecting which physical port the device is connected on (_particularly when using USB-UART adapters_).

I've tested it OK with my own F9P and M10 devices (_screenshot below is from an M10S with indoor antenna_):

<img width="357" alt="Screenshot 2024-04-09 at 13 58 02" src="https://github.com/semuconsulting/PyGPSClient/assets/28569967/947db513-89f8-4e2e-b0c3-8b6fa7823ad3">

You may want to give this a go with your own kit and let me know how you get on:

```shell
python3 -m pip install git+https://github.com/semuconsulting/PyGPSClient@RC-1.4.14
```

**Comment by GPSFan** [comment]
Still not working as it should, maybe there is a fundamental problem with how I have things set up. Are you using Linux or windows? I need to document what I'm seeing in detail, so it may take me a day of so, too many other projects...

**Reply by semuadmin** [reply]
What, specifically, is "not working as it should"?

The changes in RC 1.4.14 (_you have installed this branch, yes?_) have been tested with newer M10S and F10T devices on Windows 11, MacOS 14.4.1 (Sonoma) and Ubuntu 22.04 LTS (Jammy). My SparkFun M10S breakout board only has a single UART1 port so I'm testing it with a standard CP210n USB-UART adapter. I've duly enabled the UART1 port in the UBX `defaultport` settings as per the screen shot above (_ensure you save this setting first_) and then clicked Show Spectrum.

Here, for example, is the spectrum from a SparkFun M10S breakout board on Ubuntu 22.04:

![Screenshot from 2024-04-09 15-50-28](https://github.com/semuconsulting/PyGPSClient/assets/28569967/5e8c8b3c-5c79-4442-ab0c-c54778d7f6b6)

**NB:** I did notice one quirk on my SparkFun M10S breakout board - it would only output MON-SPAN messages on its UART1 port if the baud rate was 38400 or greater - it wouldn't work at 9600, for example. This applied on all operating systems. I'm guessing it's a manufacturer's internal Tx bandwidth constraint - I don't think it's anything to do with PyGPSClient.

**Comment by GPSFan** [comment]
As I said, I have to document just what is going on, as well as trying it with my Sparkfun M10S. I have been running the F10 at 38400 baud. I'll try at higher bauds. 
Glad you are testing on all platforms. 
Be back to you with more details as soon as I can.

**Comment by GPSFan** [comment]
Quick update, 115200 definitely has a positive effect, with some fiddling (fiddled with too many things at once) I get continuous SPAN updates.
The F10 is a dual band receiver and produces a lot more SPAN data than an M10, maybe there is some restriction as you noticed due to its increased data output.
More later.

**Comment by GPSFan** [comment]
My M10 works by just setting show spectrum in the view menu, the default baud is 38400.
For the F10 (also default baud of 38400) No spectrum when I set show spectrum. I then did a factory reset and connected to the F10 at 38400, I then set the UART1 baud to 115200 and reconnected PyGPSClient at 115200. When I set show spectrum, I get a blank spectrum display. I then go to ubx options and using  VALGET I look at the CFG_MSGOUT_UBX_MON_SPAN_UART1 RAM Layer key and I get a value of 0. Using VALSET I change that to 1 and now I get a continuously updating 2 band spectrum display.
Once I did get a spectrum after just changing the baud, re-connecting and setting show spectrum, but I have not been able to repeat that.
